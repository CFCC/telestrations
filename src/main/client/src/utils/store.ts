import { createGamekitSlice, sendEvent } from "@piticent123/gamekit-client";
import {
  configureStore,
  createAsyncThunk,
  createSlice,
  PayloadAction,
} from "@reduxjs/toolkit";
import {
  TypedUseSelectorHook,
  useSelector as useUntypedSelector,
} from "react-redux";

import { GameState, TelestrationsGame, TelestrationsPlayer } from "./types";
import _ from "lodash";
import { Settings } from "@piticent123/gamekit-client/lib/types";

const { reducer: gamekitReducer, actions: gamekitActions } =
  createGamekitSlice<Settings, TelestrationsPlayer, TelestrationsGame>();

interface AppState {
  gameState: GameState;
  activeContentId: string;
}
interface State {
  app: AppState;
  gamekit: ReturnType<typeof gamekitReducer>;
}
interface ThunkApi {
  state: State;
}

// This has to be at the top level to be throttled correctly
const _sendWritePageEvent = _.throttle((gameCode: string, content: string) => {
  sendEvent({
    route: `/app/games/${gameCode}/page/write`,
    data: content,
  });
}, 500);
export const setGuess = createAsyncThunk<void, string, ThunkApi>(
  "setGuess",
  async (content, { getState }) => {
    const {
      gamekit: { currentGame },
    } = getState();

    _sendWritePageEvent(currentGame.code, content);
  }
);

export const submitGuess = createAsyncThunk<void, string, ThunkApi>(
  "submitGuess",
  async (content, { getState }) => {
    const {
      gamekit: { currentGame },
    } = getState();

    sendEvent({
      route: `/app/games/${currentGame.code}/page/submit`,
      data: content,
    });
  }
);

export const startGame = createAsyncThunk<void, string, ThunkApi>(
  "startGame",
  async (__, { getState }) => {
    const {
      gamekit: { currentGame },
    } = getState();

    sendEvent({ route: `/games/${currentGame.code}/start` });
  }
);

export const {
  createGame,
  joinGame,
  rejoinGame,
  connectToServer,
  saveSettings,
  becomeAdmin,
} = gamekitActions;

const { actions, reducer: appReducer } = createSlice({
  name: "app",
  initialState: {
    gameState: GameState.LOGIN,
    activeContentId: "",
  } as AppState,
  reducers: {
    viewPlayerHistory: (state, action: PayloadAction<string>) => {
      state.gameState = GameState.PLAYER_HISTORY;
      state.activeContentId = action.payload;
    },
    viewNotepadHistory: (state, action: PayloadAction<string>) => {
      state.gameState = GameState.NOTEPAD_HISTORY;
      state.activeContentId = action.payload;
    },
    setGameState: (state, action: PayloadAction<GameState>) => {
      state.gameState = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(gamekitActions.rejoinGame.fulfilled, (state) => {
        state.gameState = GameState.IN_GAME;
      })
      .addCase(gamekitActions.connectToServer.fulfilled, (state) => {
        state.gameState = GameState.GAME_CODE;
      })
      .addCase(gamekitActions.joinGame.fulfilled, (state) => {
        state.gameState = GameState.WAITING_TO_START;
      })
      .addCase(submitGuess.fulfilled, (state) => {
        state.gameState = GameState.WAITING_FOR_CONTENT;
      });
  },
});

export const {
  viewNotepadHistory,
  viewPlayerHistory,
  setGameState
} = actions;

export const store = configureStore({
  reducer: { gamekit: gamekitReducer, app: appReducer },
});

export const useSelector: TypedUseSelectorHook<State> = useUntypedSelector;
