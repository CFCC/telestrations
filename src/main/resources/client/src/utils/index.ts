export function sleep(t: number): Promise<void> {
  return new Promise((res) => setTimeout(res, t));
}
