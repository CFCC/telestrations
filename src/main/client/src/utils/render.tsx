import React from "react";
import { render } from "@testing-library/react";
import { StylesProvider, ThemeProvider } from "@material-ui/core/styles";
import { Provider } from "react-redux";

import { GlobalStyles, theme } from "./theme";
import { store } from "./store";

interface WrapperProps {
  children: React.ReactNode;
}

const wrapper = ({ children }: WrapperProps) => (
  <React.Fragment>
    <StylesProvider injectFirst={true}>
      <ThemeProvider theme={theme}>
        <GlobalStyles />
        <Provider store={store}>{children}</Provider>
      </ThemeProvider>
    </StylesProvider>
  </React.Fragment>
);
const customRender = (ui: React.ReactElement) =>
  render(ui, { wrapper } as object);

export { customRender as render };
