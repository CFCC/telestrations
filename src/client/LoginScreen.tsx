import React, {useContext, useEffect} from "react";
import firebase from "firebase/app";
import * as firebaseUi from "firebaseui";
import {createGlobalStyle} from "styled-components";

import {GameContext} from "../store/client";
import {useBoolean} from "../utils/hooks";
import TitleScreen from "../components/TitleScreen";

import "firebaseui/dist/firebaseui.css"

const firebaseLoginUi = new firebaseUi.auth.AuthUI(firebase.auth());

const MakeAnonLoginABetterColor = createGlobalStyle`
    .firebaseui-idp-anonymous { 
        background-color: #000 !important;
    }
`;

export default function LoginScreen() {
    const [{user}, {setUser}] = useContext(GameContext);
    const [uiLoading,, uiShown] = useBoolean(true);
    const firebaseLoginUiContainerId = "firebaseui-auth-container";

    useEffect(() => {
        if (!user) firebaseLoginUi.start(`#${firebaseLoginUiContainerId}`, {
            callbacks: {uiShown},
            signInFlow: 'popup',
            signInOptions: [
                firebase.auth.GoogleAuthProvider.PROVIDER_ID,
                firebase.auth.PhoneAuthProvider.PROVIDER_ID,
                firebaseUi.auth.AnonymousAuthProvider.PROVIDER_ID
            ],
        });
    });

    firebase.auth().onAuthStateChanged(function(user: firebase.User | null) {
        // if (user) setUser(user);
    });

    return (
        <TitleScreen title="Log In" loading={uiLoading}>
            <MakeAnonLoginABetterColor />
            <div id={firebaseLoginUiContainerId} />
        </TitleScreen>
    );
}
