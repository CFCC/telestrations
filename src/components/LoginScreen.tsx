import React, {useEffect} from "react";
import {createGlobalStyle} from "styled-components";

import {useSelector} from "../utils/store";
import {useBoolean} from "../utils/hooks";
import TitleScreen from "./TitleScreen";
import {firebaseLoginUi, signInOptions} from "../utils/firebase";

import "firebaseui/dist/firebaseui.css";

const MakeAnonLoginABetterColor = createGlobalStyle`
    .firebaseui-idp-anonymous { 
        background-color: #000 !important;
    }
`;

export default function LoginScreen() {
    const user = useSelector(state => state.client.user);
    const [uiLoading,, uiShown] = useBoolean(true);
    const firebaseLoginUiContainerId = "firebaseui-auth-container";

    useEffect(() => {
        if (!user) firebaseLoginUi.start(`#${firebaseLoginUiContainerId}`, {
            callbacks: {uiShown},
            signInFlow: 'popup',
            signInOptions,
        });
    });

    return (
        <TitleScreen title="Log In" loading={uiLoading}>
            <MakeAnonLoginABetterColor />
            <div id={firebaseLoginUiContainerId} />
        </TitleScreen>
    );
}
