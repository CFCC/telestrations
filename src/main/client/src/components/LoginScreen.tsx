import React, { FormEvent, useEffect } from "react";
import { connectToServer, saveSettings, useSelector } from "../utils/store";
import TitleScreen from "./TitleScreen";
import { useDispatch } from "react-redux";
import {
  Button as UnstyledButton,
  IconButton,
  TextField as UnstyledTextField,
  Typography,
} from "@material-ui/core";
import { SyncOutlined as SyncIcon } from "@material-ui/icons";
import styled from "styled-components";
import { getRandomName, asImage, useAvatar } from "@piticent123/gamekit-client";
import { useInput } from "@piticent123/utils/lib/hooks"

const Form = styled.form`
  display: flex;
  flex-direction: column;
`;

const InputGroup = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-around;
  margin-bottom: 3rem;
`;

const TextField = styled(UnstyledTextField)`
  margin-right: 1rem;
`;

const Button = styled(UnstyledButton)`
  margin-bottom: 3rem;
`;

export default function LoginScreen() {
  const {
    id,
    name: defaultName,
    avatar: defaultAvatar,
  } = useSelector((state) => state.gamekit.settings);
  const dispatch = useDispatch();
  const [avatar, randomizeImage] = useAvatar();
  const [name, setName] = useInput(defaultName ?? getRandomName());

  useEffect(() => {
    if (id && defaultName && defaultAvatar) {
      dispatch(connectToServer());
    }
  }, []);

  async function handleGoToLobby(e: FormEvent<HTMLFormElement>) {
    e.preventDefault();

    await dispatch(saveSettings({ name, avatar }));
    await dispatch(connectToServer());
  }

  function randomizeName() {
    setName(getRandomName());
  }

  return (
    <TitleScreen title="Log In">
      <Form onSubmit={handleGoToLobby}>
        <Typography variant="caption" color="textSecondary">
          Profile Picture:
        </Typography>
        <InputGroup>
          <img src={asImage(avatar)} alt="Avatar" />
          <IconButton onClick={randomizeImage} aria-label="Randomize Image">
            <SyncIcon />
          </IconButton>
        </InputGroup>
        <InputGroup>
          <TextField
            value={name}
            onChange={setName}
            placeholder="Name"
            label="Name"
            variant="outlined"
          />
          <IconButton onClick={randomizeName} aria-label="Randomize Name">
            <SyncIcon />
          </IconButton>
        </InputGroup>
        <Button type="submit" color="primary" variant="contained">
          Go to Lobby
        </Button>
      </Form>
    </TitleScreen>
  );
}
