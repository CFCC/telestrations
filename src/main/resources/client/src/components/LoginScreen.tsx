import React, { FormEvent, useEffect, useState } from "react";
import { actions, goToLobby, saveSettings, useSelector } from "../utils/store";
import { useInput } from "../utils/hooks";
import TitleScreen from "./TitleScreen";
import { createAvatar } from "@dicebear/avatars";
import * as sprites from "@dicebear/avatars-human-sprites";
import { useDispatch } from "react-redux";
import { v4 as uuidv4 } from "uuid";
import { animals, colors, uniqueNamesGenerator } from "unique-names-generator";
import {
  Button as UnstyledButton,
  IconButton,
  TextField as UnstyledTextField,
  Typography,
} from "@material-ui/core";
import { SyncOutlined as SyncIcon } from "@material-ui/icons";
import styled from "styled-components";

const Form = styled.form`
  display: flex;
  flex-direction: column;
`;

const InputGroup = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-around;
  margin-bottom: 3rem;
`;

const TextField = styled(UnstyledTextField)`
  margin-right: 1rem;
`;

const Button = styled(UnstyledButton)`
  margin-bottom: 3rem;
`;

const getRandomName = () =>
  uniqueNamesGenerator({
    dictionaries: [colors, animals],
    length: 2,
    separator: " ",
    style: "capital",
  });

export default function LoginScreen() {
  const {
    id,
    name: defaultName,
    avatar: defaultAvatar,
  } = useSelector((state) => state.settings);
  const dispatch = useDispatch();
  const [avatar, setAvatar] = useState(defaultAvatar ?? uuidv4());
  const [name, setName] = useInput(defaultName ?? getRandomName());

  useEffect(() => {
    if (id && defaultName && defaultAvatar) {
      dispatch(goToLobby());
    }
  }, []);

  async function handleGoToLobby(e: FormEvent<HTMLFormElement>) {
    e.preventDefault();

    await dispatch(saveSettings({ name, avatar, id: id || uuidv4() }));
    await dispatch(goToLobby());
  }

  function randomizeName() {
    setName(getRandomName());
  }

  function randomizeImage() {
    setAvatar(uuidv4());
  }

  return (
    <TitleScreen title="Log In">
      <Form onSubmit={handleGoToLobby}>
        <Typography variant="caption" color="textSecondary">
          Profile Picture:
        </Typography>
        <InputGroup>
          <img
            src={createAvatar(sprites, {
              seed: avatar,
              width: 150,
              height: 150,
              dataUri: true,
            })}
            alt="Avatar"
          />
          <IconButton onClick={randomizeImage} aria-label="Randomize Image">
            <SyncIcon />
          </IconButton>
        </InputGroup>
        <InputGroup>
          <TextField
            value={name}
            onChange={setName}
            placeholder="Name"
            label="Name"
            variant="outlined"
          />
          <IconButton onClick={randomizeName} aria-label="Randomize Name">
            <SyncIcon />
          </IconButton>
        </InputGroup>
        <Button type="submit" color="primary" variant="contained">
          Go to Lobby
        </Button>
      </Form>
    </TitleScreen>
  );
}
