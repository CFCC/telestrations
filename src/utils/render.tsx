import React from "react";
import {render} from "@testing-library/react";
import {StylesProvider, ThemeProvider} from "@material-ui/core/styles";
import {Provider} from "react-redux";

import {GlobalStyles, theme} from "./theme";
import {store} from "../store/client";

interface WrapperProps {
    children: React.ReactNode;
}

const wrapper = (store: any) => ({children}: WrapperProps) => (
    <React.Fragment>
        <StylesProvider injectFirst={true}>
            <ThemeProvider theme={theme}>
                <GlobalStyles />
                <Provider store={store}>
                    {children}
                </Provider>
            </ThemeProvider>
        </StylesProvider>
    </React.Fragment>
);
const customRender = (ui: React.ReactElement, options: any) => render(ui, {wrapper, ...options});

export {customRender as render};
