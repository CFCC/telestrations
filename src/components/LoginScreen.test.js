import {render} from "../utils/render";
import {Standard} from "./LoginScreen.stories";

jest.mock("../utils/firebase");

it('Renders', async () => {
    const {asFragment} = render(Standard());
    expect(asFragment()).toMatchSnapshot();
});
