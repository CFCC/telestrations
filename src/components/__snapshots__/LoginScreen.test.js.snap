// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm kwsLvR"
  >
    <img
      alt="Telestrations"
      class="sc-AxiKw iXGiQC"
      src="/logo.png"
    />
    <p
      class="MuiTypography-root sc-AxhCb bDZFln MuiTypography-body1"
    >
      Log In
    </p>
    <div
      id="firebaseui-auth-container"
    />
    <div
      class="MuiCircularProgress-root sc-AxirZ eovyrP MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
      role="progressbar"
      style="width: 40px; height: 40px;"
    >
      <svg
        class="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
